EIPS/eip-3009.md

```diff
@@ -0,0 +1,472 @@
+      { name: "name", type: "string" },
+      { name: "version", type: "string" },
+      { name: "chainId", type: "uint256" },
+      { name: "verifyingContract", type: "address" },
+    ],
+    TransferWithAuthorization: [
+      { name: "from", type: "address" },
+      { name: "to", type: "address" },
+      { name: "value", type: "uint256" },
+      { name: "validAfter", type: "uint256" },
+      { name: "validBefore", type: "uint256" },
+      { name: "nonce", type: "bytes32" },
+    ],
+  },
+  domain: {
+    name: tokenName,
+    version: tokenVersion,
+    chainId: selectedChainId,
+    verifyingContract: tokenAddress,
+  },
+  primaryType: "TransferWithAuthorization",
+  message: {
+    from: userAddress,
+    to: recipientAddress,
+    value: amountBN.toString(10),
+    validAfter: 0,
+    validBefore: Math.floor(Date.now() / 1000) + 3600, // Valid for an hour
+    nonce: Web3.utils.randomHex(32),
+  },
+};
+
+const signature = await ethereum.request({
+  method: "eth_signTypedData_v4",
+  params: [userAddress, JSON.stringify(data)],
+});
+
+const v = "0x" + signature.slice(130, 132);
+const r = signature.slice(0, 66);
+const s = "0x" + signature.slice(66, 130);
+```
+
+## Rationale
+
+### Unique Random Nonce, Instead of Sequential Nonce
+
+One might say transaction ordering is one reason why sequential nonces are preferred. However, sequential nonces do not actually help achieve transaction ordering for meta transactions in practice:
+
+- For native Ethereum transactions, when a transaction with a nonce value that is too-high is submitted to the network, it will stay pending until the transactions consuming the lower unused nonces are confirmed.
+- However, for meta-transactions, when a transaction containing a sequential nonce value that is too high is submitted, instead of staying pending, it will revert and fail immediately, resulting in wasted gas.
+- The fact that miners can also reorder transactions and include them in the block in the order they want (assuming each transaction was submitted to the network by different meta-transaction relayers) also makes it possible for the meta-transactions to fail even if the nonces used were correct. (e.g. User submits nonces 3, 4 and 5, but miner ends up including them in the block as 4,5,3, resulting in only 3 succeeding)
+- Lastly, when using different applications simultaneously, in absence of some sort of an off-chain nonce-tracker, it is not possible to determine what the correct next nonce value is if there exists nonces that are used but haven't been submitted and confirmed by the network.
+- This is especially problematic now that high gas price fluctuations often cause transactions to "get stuck" in the pool for a long time. Today, if you make a meta-transaction that uses a sequential nonce from one app, and switch to another app to make another meta-transaction before the previous one confirms, the same nonce will be used twice, resulting in one of the transactions failing.
```

Comment:
> now that high gas price fluctuations often cause transactions to "get stuck" in the pool for a long time

EIPs should be worded for a target audience reading it in 5+ years.  Since we have no idea what gas prices will be like at that time, we shouldn't discuss their state (e.g., volatility) as an invariant or really even mention anything about the "current" situation.